File: server/src/models/RoomPlayStatus.ts
import { Model, DataTypes } from 'sequelize';
import sequelize from '../db/connection';
import Room from './Room';

class RoomPlayStatus extends Model {
    declare id: number;
    declare roomId: number;
    declare paused: boolean;    // true if the video is paused
    declare time: number;   // current time in seconds
    declare timestamp: number;  // timestamp of the last user action, in milliseconds
    declare videoId: number;    // the video id
}

RoomPlayStatus.init({
    id: {
        type: DataTypes.INTEGER,
        primaryKey: true,
        autoIncrement: true
    },
    roomId: {
        type: DataTypes.INTEGER,
        allowNull: false,
        references: {
            model: Room,
            key: 'id'
        }
    },
    paused: {
        type: DataTypes.BOOLEAN,
        defaultValue: false
    },
    time: {
        type: DataTypes.DOUBLE,
        defaultValue: 0
    },
    timestamp: {
        type: DataTypes.BIGINT,
        defaultValue: 0
    },
    videoId: {
        type: DataTypes.INTEGER,
        defaultValue: 0
    }
}, {
    sequelize,
    modelName: 'RoomPlayStatus',
    tableName: 'room_play_status',
    timestamps: false
});

export default RoomPlayStatus;

File: server/internal/models/room_play_status.go
package models

import (
	"gorm.io/gorm"
)

// RoomPlayStatus represents the current playback status of a room
type RoomPlayStatus struct {
	ID        uint           `gorm:"primaryKey;autoIncrement" json:"id"`
	RoomID    uint           `gorm:"not null;uniqueIndex" json:"roomId"`
	Paused    bool           `gorm:"default:false" json:"paused"`
	Time      float64        `gorm:"default:0" json:"time"`
	Timestamp int64          `gorm:"default:0" json:"timestamp"`
	VideoID   uint           `gorm:"default:0" json:"videoId"`
	DeletedAt gorm.DeletedAt `gorm:"index" json:"-"`

	// Associations
	Room *Room `gorm:"foreignKey:RoomID" json:"room,omitempty"`
}

// TableName specifies the table name for RoomPlayStatus model
func (RoomPlayStatus) TableName() string {
	return "room_play_status"
}
